### ISSUE TITLE
Continuous Integration for Tests

### ISSUE THREAD

**Issue by truell20** [jesserayadkins/lily#172]

**Comment by jesserayadkins** [comment]
I'd like to have this eventually. Right now though, as I'm the only developer, I've set up a pre-commit script that links to the pre-commit-hook.py script and runs it to make sure all tests pass.

**Comment by TyRoXx** [comment]
In my fork where I currently port lily to Visual C++ 2015 I use Travis to make sure I don't break Linux too hard: https://github.com/TyRoXx/lily/blob/34ae63e0af83bca83156d47810a7ba3c639e8735/.travis.yml

I also use Appveyor, but I have not gotten ``pre-commit-hook.py`` to work on Windows yet. I can run lily.exe tests individually and they succeed usually, but with Python the child process just prints some generic "file not found" Windows error message to stderr.

**Comment by jesserayadkins** [comment]
Wow! I've been wanting Lily to be ported to Windows for a while now, but simply haven't gotten around to it. That alone is going to be a great help! Truth be told, I can't remember if I ever tested Lily on Windows, so I'm not surprised that the `pre-commit-hook.py` script is failing.

I'm currently in the process of setting up a development environment on my Windows machine. If I run into the same problem and can figure it out, then I'll send you a PR. You mention that Lily tests pass 'usually', and I'd like to get to the bottom of that as well.

**Comment by TyRoXx** [comment]
The tests are fine, they to not fail spuriously or something. There was only one failure due to a different behaviour of the VC++ compiler.

I had to change this:
```C
#define lily_move_double(target, raw) \
lily_move(target, (lily_raw_value)raw, VAL_IS_DOUBLE)
```

Into this:
```C
#define lily_move_double(target, raw) \
lily_move(target, (lily_raw_value){.doubleval = raw}, VAL_IS_DOUBLE)
```

Otherwise the value would be wrong. I did not analyze this issue further as my quick fix made the test pass. It looks like VC++ cast the double into an integer (the first field of the union) while GCC seems to find the correct field to initialize.

**Comment by jesserayadkins** [comment]
After some trial and error, I managed to get the `pre-commit-hook.py` script to work. It turns out that `shell=True` was causing the majority of problems. The rest was a lot of `os.sep` needing to be peppered in the right places. I've pushed a commit wherein the script runs on my Windows machine and on *nix box too.

Unfortunately, running the script on Windows is troublesome. About a dozen tests are crashing, and one (an importing test) is hanging. My next move is to start combing through the tests and smash any invalid reads/writes/etc. Hopefully fixing those will get the Windows tests to pass.

I've pushed the fix, along with some other fixes. Hopefully you'll be able to merge them into your fork smoothly (I assume that's how it works, but let me know if there's something else I should have done instead).
